### ISSUE TITLE
featurelocation contains code if not all brackets are closed

### ISSUE THREAD

**Issue by zottelchin** [clhunsen/cppstats#22]
I'm using cppstats to get the feature location in c-code. The resulting csv-file is parsed by an python script to save the data into a database. When the script failed to run, i was looking into the csv-file to determine the source of the error.

The c-file processed was from the openldap repo and contained a ifdef statement with an opening bracket and no closing bracket. Resulting in an output of the code following the statement (including line-breaks what lead to an invalid csv format). The line in question was https://git.openldap.org/openldap/openldap/-/blob/86c216d10cad596eb8ca37bd9263f56b8dc1042a/libraries/libldap/os-ip.c#L617

I think it would be a good idea to not only end an entry on the right closing bracket, but also when the end of the line is reached.

**Comment by xai** [comment]
Nice catch!

Can you point us to a standard defining this (just using opening parenthesis) as valid syntax? I am not aware of any such standard existing.
If there is indeed no such standard, I would rather expect a warning or even an error, as this might as well be the result of a broken merge or something alike. Semantically interpreting things that are syntactically incorrect is *really* hard and probably out of scope for cppstats, but @clhunsen will know best in this case.

**Comment by zottelchin** [comment]
I have not found a standard for the parentheses. However, it is defined in the documentation of cpp that a marco ends with the end of the line.

> The macro’s body ends at the end of the ‘#define’ line. You may continue the definition onto multiple lines, if necessary, using backslash-newline. When the macro is expanded, however, it will all come out on one line.

https://gcc.gnu.org/onlinedocs/cpp/Object-like-Macros.html#Object-like-Macros

**Comment by clhunsen** [comment]
Hi @zottelchin,

sorry for replying late, I am not working at the university anymore and looking after cppstats is not a priority for me nowadays.

Anyway, I have looked into the problem. Here are my two cents to add for the reported issue:
- Based on the definitions of `cpp` by [Kernighan and Ritchie](https://edutechlearners.com/the-c-programming-language-by-kernighan-ritchie-pdf/) (see Appendix A.12.5), `#if defined` either contains two matching parentheses or none at all. Nevertheless, in my opinion, the line break should end the expression. But...
- In the end, the reported issue arises from the source-code translation into srcML using the `srcml` binary (see the resulting file [os-ip.xml](https://github.com/clhunsen/cppstats/files/10203324/os-ip.xml.txt), but also https://github.com/srcML/srcML/issues/1874). Unfortunately, to this end, we cannot do anything about the problem within cppstats as we rely on the srcML structure.

My recommendation is twofold: Either correct the grammar error or exclude the file from analysis. In both cases, you should note the file as being erroneous and mention that fact in any analysis descriptions.

tl;dr: The error occurs during srcML translation such that we cannot do about it in cppstats. I will close here.
